src/lib/support/BufferReader.h

```diff
@@ -0,0 +1,146 @@
+/*
+ *
+ *    Copyright (c) 2020 Project CHIP Authors
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+/**
+ *  @file
+ *    Utility classes for safely reading from size-limited buffers.
+ */
+
+#pragma once
+
+#include <climits>
+#include <lib/core/CHIPEncoding.h>
+#include <lib/core/CHIPError.h>
+#include <lib/support/CodeUtils.h>
+#include <stdint.h>
+
+namespace chip {
+
+/**
+ *  @class LittleEndianReader
+ *
+ *  Simple reader for reading little-endian things out of buffers.
+ */
+class LittleEndianReader
+{
+public:
+    /**
+     * Create a data model reader from a given buffer and length.
+     *
+     * @param buffer The octet buffer to read from.  The caller must ensure
+     *               (most simply by allocating the reader on the stack) that
+     *               the buffer outlives the reader.  The buffer is allowed to
+     *               be null if buf_len is 0.
+     * @param buf_len The number of octets in the buffer.
+     */
+    LittleEndianReader(const uint8_t * buffer, uint16_t buf_len) : mBufStart(buffer), mReadPtr(buffer), mAvailable(buf_len) {}
+
+    /**
+     * Number of octets we have read so far.  This might be able to go away once
+     * we do less switching back and forth between DataModelReader and raw
+     * buffers.
+     */
+    uint16_t OctetsRead() const { return static_cast<uint16_t>(mReadPtr - mBufStart); }
+
+    /**
+     * Read a single octet.
+     *
+     * @param [out] octet Where the octet goes.
+     *
+     * @return Whether the read succeeded.  The read can fail if there are not
+     *         enough octets available.
+     */
+    CHECK_RETURN_VALUE
+    CHIP_ERROR ReadOctet(uint8_t * octet) { return Read(octet); }
```

Comment:
Why not `Read8`? This seems to be an odd API nomenclature exception.

Reply:
Good catch.  This started out as a different class with more-semantic function names, and I forgot to rename this one.  Will fix.